### ISSUE TITLE
hard crash when plotting large graph with layout "kk"

### ISSUE THREAD

**Issue by mkoohafkan** [igraph/rigraph#348]
R crashes hard when trying to plot a large, dense graph using layout `'kk'` (process terminates without warning or error). It would be preferable if it could fail gracefully.

```r
set.seed(50)

library(igraph)

nodes = data.frame(name = paste0("n", 1:150000))

edges = unique(data.frame(
  from = sample(nodes$name, 150000, replace = TRUE),
  to = sample(nodes$name, 150000, replace = TRUE)
))

g = graph_from_data_frame(edges, FALSE, vertices = nodes)

plot(g, layout=layout_with_kk)
```

Session Info:

```r
## R version 3.6.1 (2019-07-05)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 17763)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United States.1252 
## [2] LC_CTYPE=English_United States.1252   
## [3] LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] igraph_1.2.4.1
## 
## loaded via a namespace (and not attached):
## [1] compiler_3.6.1  magrittr_1.5    pkgconfig_2.0.3
```

**Comment by szhorvat** [comment]
No doubt this happens because `igraph_layout_kamada_kawai` allocates several matrices of size `n*n` where `n` is the number of nodes. These types of issues will be addressed in version 0.10 of the C core.

**Comment by ntamas** [comment]
Even if we address it in 0.10, the Kamada-Kawai layout is not designed for and will not be usable for graphs of this size. There are better alternatives like the DrL layout, although I'm pretty sure that for 150K vertices you will end up with a hairball anyway.

**Comment by szhorvat** [comment]
By addressing it, I meant including a check to make sure we're not trying to exceed the max matrix size. The improvement would that that there's an error message instead of a crash.

**Comment by ntamas** [comment]
As for the termination, I have a feeling that error handling is somehow broken in R-igraph, at least for out-of-memory conditions. When we allocate a matrix in igraph's C core and it turns out to be too big, we return an `IGRAPH_ENOMEM` error code, and this should be handled by R. It looks like this does not happen.

**Comment by szhorvat** [comment]
Shall we limit `kk` to no more than 10000 or maybe 20000 vertices? That would 2.2 GB or 9 GB of memory. The computation probably won't finish in reasonable time even for 10000. BTW `kk` is not interruptible ... should fix that as well.

**Comment by ntamas** [comment]
Interruption support should be implemented, yes. As for limiting, I'm not very fond of that. Basically, it seems like we cannot rely on `malloc()` returning `NULL` if there is not enough memory available, at least not on Linux when overcommitting is enabled in the kernel - but we cannot put guards around every single use of `malloc()` either, so fixing it in one place in the KK layout probably wouldn't bring much benefit to the project overall.

I wonder whether we should add support for a progress bar, though. In theory, the KK layout has a maximum iteration count and an epsilon value, so it will never perform more than `maxiter` iterations -- but it may also bail out earlier, in which case the progress bar may provide a false estimate of the remaining time.

**Comment by szhorvat** [comment]
> Interruption support should be implemented, yes.

I'll look into that.

> As for limiting, I'm not very fond of that.

Agreed, I didn't love the idea either. BTW for a 10000-vertex connected preferential attachment graph the algorithm finished in 2 minutes on my computer. That's not very long at all.

**Comment by szhorvat** [comment]
Closing in favour of #405, they will be solved together in the C core for 2.0.
